src/RootHitFiles.jl

```diff
@@ -0,0 +1,71 @@
+"""
+    RootHitFile(file::Union{IOStream, AbstractString})
+
+represents a `.root.hits` file, which can be iterated or read to yield events, containg data
+on hits.
+"""
+struct RootHitFile
+    stream::IOBuffer
+end
+
+RootHitFile(stream::IOStream) = RootHitFile(IOBuffer(mmap(stream)))
```

Comment:
Does mmap actually improve performance here?

Reply:
I tried a bunch of different things and memory mapped IOBuffer was the best (by quite a lot) when I last ran benchmarks. But now that you ask I'm curious again, so I might see if that's changed.

Reply:
Ok here are the comparisons for using a straight up `IOStream` vs a `IOBuffer` of a memory mapped file.

## TLDR

Time to collect the entire file goes from 6.254 ms without mmap to 0.2672 ms with mmap.

Time to read first event goes from 644.408 μs without mmap to 41.231 μs.

Time improvement comes at the cost of about 18% more total memory allocation.

## Benchmarks

I used the following three benchmarks:

```julia
SUITE = BenchmarkGroup()

SUITE["construct"] = @benchmarkable RootHitFile($eventpath)

SUITE["read"] = @benchmarkable read(RootHitFile($eventpath))

SUITE["collect"] = @benchmarkable collect(RootHitFile($eventpath))
```

### IOStream (no memory mapping)

```julia
julia> r = run(SUITE)
3-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "construct" => Trial(9.499 μs)
  "collect" => Trial(6.254 ms)
  "read" => Trial(644.408 μs)

julia> r["collect"]
BenchmarkTools.Trial: 
  memory estimate:  203.45 KiB
  allocs estimate:  2444
  --------------
  minimum time:     6.254 ms (0.00% GC)
  median time:      6.743 ms (0.00% GC)
  mean time:        6.806 ms (0.25% GC)
  maximum time:     27.951 ms (0.00% GC)
  --------------
  samples:          735
  evals/sample:     1

julia> r["read"]
BenchmarkTools.Trial: 
  memory estimate:  21.48 KiB
  allocs estimate:  261
  --------------
  minimum time:     644.408 μs (0.00% GC)
  median time:      683.241 μs (0.00% GC)
  mean time:        698.904 μs (0.34% GC)
  maximum time:     15.795 ms (37.35% GC)
  --------------
  samples:          7145
  evals/sample:     1
```

### IOBuffer (memory mapped)

```julia
julia> r = run(SUITE)
3-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "construct" => Trial(14.125 μs)
  "collect" => Trial(267.160 μs)
  "read" => Trial(41.231 μs)

julia> r["collect"]
BenchmarkTools.Trial: 
  memory estimate:  239.48 KiB
  allocs estimate:  2435
  --------------
  minimum time:     267.160 μs (0.00% GC)
  median time:      274.981 μs (0.00% GC)
  mean time:        302.123 μs (3.58% GC)
  maximum time:     19.460 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> r["read"]
BenchmarkTools.Trial: 
  memory estimate:  25.64 KiB
  allocs estimate:  266
  --------------
  minimum time:     41.231 μs (0.00% GC)
  median time:      43.882 μs (0.00% GC)
  mean time:        52.881 μs (3.83% GC)
  maximum time:     15.698 ms (27.84% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

Reply:
Could you include BufferedStreams.jl in the comparison (with and without mmap)? That could be valuable for similar ASCII-handling cases.

Reply:
I've been using it here already: https://github.com/legend-exp/LegendTextIO.jl/blob/268cb42a1cf6f80ca429d74fc0bf11ca72adc515/src/geant4_csv.jl#L14

Reply:
Will do